spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/ApiKeyAuthTokenVerifier.java

```diff
@@ -20,166 +20,124 @@
+    private final ApiKeyFetcher apiKeyFetcher;
+    private final LocalApiKeyVerifier localApiKeyVerifier;
 
-    @Value("${ygg.admin.resources.search-service-by-api-key-value}")
-    private String adminResourceToGetServiceByApiKeyValue;
+    public ApiKeyAuthTokenVerifier(LocalApiKeyManager localApiKeyManager,
+                                   ApiKeyFetcher apiKeyFetcher, LocalApiKeyVerifier localApiKeyVerifier) {
+        this.localApiKeyManager = localApiKeyManager;
+        this.apiKeyFetcher = apiKeyFetcher;
+        this.localApiKeyVerifier = localApiKeyVerifier;
+    }
 
     @Override
     public Authentication authenticate(Authentication authentication) throws AuthenticationException {
-
         try {
-            Assert.notNull(authentication.getCredentials());
-            Assert.isInstanceOf(String.class, authentication.getCredentials());
-            String apiKey = (String) authentication.getCredentials();
-            Assert.hasText(apiKey);
-
-            //TODO this is just a workaround for now: check for fixed API key to allow CoatRack admin access
-            if (apiKey.equals(ApiKey.API_KEY_FOR_YGG_ADMIN_TO_ACCESS_PROXIES)) {
-
-                Set<SimpleGrantedAuthority> authoritiesGrantedToYggAdmin = new HashSet<>();
-                authoritiesGrantedToYggAdmin.add(new SimpleGrantedAuthority(ServiceApiAccessRightsVoter.ACCESS_SERVICE_AUTHORITY_PREFIX + "refresh"));
-
-                ApiKeyAuthToken apiKeyAuthToken = new ApiKeyAuthToken(apiKey, authoritiesGrantedToYggAdmin);
-                apiKeyAuthToken.setAuthenticated(true);
-                return apiKeyAuthToken;
-            }
-
-            // regular api consumer's api key check
-            if (isApiKeyValid(apiKey)) {
-                // key is valid, now get service api URI identifier
-                ServiceApi serviceApi = getServiceApiByApiKey(apiKey);
-                String uriIdentifier = serviceApi.getUriIdentifier();
-
-                // add uri identifier as granted authority
-                ApiKeyAuthToken apiKeyAuthToken = new ApiKeyAuthToken(apiKey, Collections.singleton(
-                        new SimpleGrantedAuthority(
-                                ServiceApiAccessRightsVoter.ACCESS_SERVICE_AUTHORITY_PREFIX + uriIdentifier)));
-
-                apiKeyAuthToken.setAuthenticated(true);
-                return apiKeyAuthToken;
-            } else {
-                // TODO log this via MetricsCounterService?
-                throw new BadCredentialsException("API key value is not valid");
-            }
-        } catch (IllegalArgumentException e) {
-            log.debug("API key parameter missing or invalid: " + e.getMessage());
-            throw new BadCredentialsException("API key parameter missing or invalid: " + e.getMessage());
+            log.debug("Verifying the authentication {}.", authentication.getName());
+            return createApiKeyAuthToken(authentication);
```

Comment:
This is hard to understand: strictly speaking, this method now consists of one line, which calls a method with the same signature that this method has. Seems unnecessarily complicated. Why not put the main logic of the authentication into this method?

But if you want to keep it that way, then the called method should also be named similar to `authenticate`, e.g. `authenticateDelegated` or `authenticateInternal`.

Reply:
My thought process was the following: A method should do exactly one thing at a time. Exception handling and token creation logic are two different things. I should not have to read any exception handling if I am just interested in learning, how the token creation logic works and vice versa.

I found the name "authenticate" hard to understand when I saw it the first time as it could mean a lot of things. Creating an authentication token is exactly what the function does. Isn't "createApiKeyAuthToken" therefore more descriptive?

Reply:
We should discuss this again, having in mind the implemented interface method:
https://docs.spring.io/spring-security/site/docs/4.2.20.RELEASE/apidocs/org/springframework/security/authentication/AuthenticationManager.html#authenticate-org.springframework.security.core.Authentication-